test/solve_dense.cpp

```diff
@@ -53,26 +53,114 @@ void solveTester(const int m, const int n, const int k, double eps)
+
+        X1 = X;
+    } else {
+        //! [ex_solve_lower]
+        af::array X = af::solve(AT, B0, AF_MAT_LOWER);
+        //! [ex_solve_lower]
+
+        X1 = X;
+    }
+
+    af::array B1 = af::matmul(AT, X1);
+
+    ASSERT_NEAR(0, af::sum<double>(af::abs(real(B0 - B1))) / (n * k), eps);
+    ASSERT_NEAR(0, af::sum<double>(af::abs(imag(B0 - B1))) / (n * k), eps);
+}
+
+#define SOLVE_TESTS(T, eps)                             \
+    TEST(SOLVE_LU, T##Reg)                              \
+    {                                                   \
+        solveLUTester<T>(1000, 100, eps);               \
+    }                                                   \
+    TEST(SOLVE_LU, T##RegMultiple)                      \
+    {                                                   \
+        solveLUTester<T>(2048, 512, eps);               \
+    }                                                   \
+    TEST(SOLVE_Upper, T##Reg)                           \
+    {                                                   \
+        solveTriangleTester<T>(1000, 100, true, eps);   \
+    }                                                   \
+    TEST(SOLVE_Upper, T##RegMultiple)                   \
+    {                                                   \
+        solveTriangleTester<T>(2048, 512, true, eps);   \
+    }                                                   \
+    TEST(SOLVE_Lower, T##Reg)                           \
+    {                                                   \
+        solveTriangleTester<T>(1000, 100, false, eps);  \
+    }                                                   \
+    TEST(SOLVE_Lower, T##RegMultiple)                   \
+    {                                                   \
+        solveTriangleTester<T>(2048, 512, false, eps);  \
+    }                                                   \
+    TEST(SOLVE, T##Square)                              \
+    {                                                   \
+        solveTester<T>(1000, 1000, 100, eps);           \
+    }                                                   \
+    TEST(SOLVE, T##SquareMultiple)                      \
+    {                                                   \
+        solveTester<T>(2048, 2048, 512, eps);           \
+    }                                                   \
+    TEST(SOLVE, T##RectUnder)                           \
+    {                                                   \
+        solveTester<T>(800, 1000, 200, eps);            \
+    }                                                   \
+    TEST(SOLVE, T##RectUnderMultiple)                   \
+    {                                                   \
+        solveTester<T>(1536, 2048, 400, eps);           \
+    }                                                   \
+    TEST(SOLVE, T##RectOverMultiple)                    \
+    {                                                   \
+        solveTester<T>(1536, 1024, 1, eps);             \
```

Comment:
These tests are too large. make smaller tests. I will try to have a different target for large tests.